# importing the data set 

In [1]:
import pandas as pd

df = pd.read_csv("diabetic_data.csv")

print("Rows:", df.shape[0])
print("Columns:", df.shape[1])

df.head()

Rows: 101766
Columns: 50


,encounter_id,patient_nbr,race,gender,age,weight,admission_type_id,discharge_disposition_id,admission_source_id,time_in_hospital,...,citoglipton,insulin,glyburide-metformin,glipizide-metformin,glimepiride-pioglitazone,metformin-rosiglitazone,metformin-pioglitazone,change,diabetesMed,readmitted
0,2278392,8222157,Caucasian,Female,[0-10),?,6,25,1,1,...,No,No,No,No,No,No,No,No,No,NO
1,149190,55629189,Caucasian,Female,[10-20),?,1,1,7,3,...,No,Up,No,No,No,No,No,Ch,Yes,>30
2,64410,86047875,AfricanAmerican,Female,[20-30),?,1,1,7,2,...,No,No,No,No,No,No,No,No,Yes,NO
3,500364,82442376,Caucasian,Male,[30-40),?,1,1,7,2,...,No,Up,No,No,No,No,No,Ch,Yes,NO
4,16680,42519267,Caucasian,Male,[40-50),?,1,1,7,1,...,No,Steady,No,No,No,No,No,Ch,Yes,NO


### identified in the data set 101766 rows and 50 columns with different attribute values 

In [2]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 101766 entries, 0 to 101765
Data columns (total 50 columns):
 #   Column                    Non-Null Count   Dtype 
---  ------                    --------------   ----- 
 0   encounter_id              101766 non-null  int64 
 1   patient_nbr               101766 non-null  int64 
 2   race                      101766 non-null  object
 3   gender                    101766 non-null  object
 4   age                       101766 non-null  object
 5   weight                    101766 non-null  object
 6   admission_type_id         101766 non-null  int64 
 7   discharge_disposition_id  101766 non-null  int64 
 8   admission_source_id       101766 non-null  int64 
 9   time_in_hospital          101766 non-null  int64 
 10  payer_code                101766 non-null  object
 11  medical_specialty         101766 non-null  object
 12  num_lab_procedures        101766 non-null  int64 
 13  num_procedures            101766 non-null  int64 
 14  num_

In [3]:
df.columns.tolist()

['encounter_id',
 'patient_nbr',
 'race',
 'gender',
 'age',
 'weight',
 'admission_type_id',
 'discharge_disposition_id',
 'admission_source_id',
 'time_in_hospital',
 'payer_code',
 'medical_specialty',
 'num_lab_procedures',
 'num_procedures',
 'num_medications',
 'number_outpatient',
 'number_emergency',
 'number_inpatient',
 'diag_1',
 'diag_2',
 'diag_3',
 'number_diagnoses',
 'max_glu_serum',
 'A1Cresult',
 'metformin',
 'repaglinide',
 'nateglinide',
 'chlorpropamide',
 'glimepiride',
 'acetohexamide',
 'glipizide',
 'glyburide',
 'tolbutamide',
 'pioglitazone',
 'rosiglitazone',
 'acarbose',
 'miglitol',
 'troglitazone',
 'tolazamide',
 'examide',
 'citoglipton',
 'insulin',
 'glyburide-metformin',
 'glipizide-metformin',
 'glimepiride-pioglitazone',
 'metformin-rosiglitazone',
 'metformin-pioglitazone',
 'change',
 'diabetesMed',
 'readmitted']

## checking the missing values 

In [4]:
df.isnull().sum().sort_values(ascending=False)

max_glu_serum               96420
A1Cresult                   84748
race                            0
gender                          0
age                             0
weight                          0
admission_type_id               0
discharge_disposition_id        0
admission_source_id             0
time_in_hospital                0
payer_code                      0
medical_specialty               0
num_lab_procedures              0
num_procedures                  0
num_medications                 0
number_outpatient               0
encounter_id                    0
patient_nbr                     0
number_inpatient                0
number_emergency                0
diag_1                          0
diag_2                          0
number_diagnoses                0
diag_3                          0
metformin                       0
repaglinide                     0
nateglinide                     0
chlorpropamide                  0
glimepiride                     0
acetohexamide 

#  Target Variable

In [5]:
df['readmitted'].value_counts()

readmitted
NO     54864
>30    35545
<30    11357
Name: count, dtype: int64

In [6]:
df['readmitted'].value_counts(normalize=True) * 100

readmitted
NO     53.911916
>30    34.928169
<30    11.159916
Name: proportion, dtype: float64

In [7]:
df['readmitted'].unique()

array(['NO', '>30', '<30'], dtype=object)

## Check the hidden missing values

In [8]:
(df == '?').sum().sort_values(ascending=False)

weight                      98569
medical_specialty           49949
payer_code                  40256
race                         2273
diag_3                       1423
diag_2                        358
diag_1                         21
admission_type_id               0
patient_nbr                     0
encounter_id                    0
time_in_hospital                0
admission_source_id             0
num_lab_procedures              0
num_procedures                  0
num_medications                 0
discharge_disposition_id        0
gender                          0
age                             0
number_inpatient                0
number_emergency                0
number_outpatient               0
number_diagnoses                0
max_glu_serum                   0
A1Cresult                       0
metformin                       0
repaglinide                     0
nateglinide                     0
chlorpropamide                  0
glimepiride                     0
acetohexamide 

## Checking duplicates

In [9]:
df.duplicated().sum()

np.int64(0)

In [10]:
print("Unique encounters:", df['encounter_id'].nunique())
print("Unique patients:", df['patient_nbr'].nunique())

Unique encounters: 101766
Unique patients: 71518


In [11]:
df['readmitted'].value_counts()

readmitted
NO     54864
>30    35545
<30    11357
Name: count, dtype: int64

In [12]:
df['readmitted'].value_counts(normalize=True) * 100

readmitted
NO     53.911916
>30    34.928169
<30    11.159916
Name: proportion, dtype: float64

In [13]:
df['readmitted'].unique()

array(['NO', '>30', '<30'], dtype=object)

* This is important: 101,766 encounters ≠ 101,766 patients.

A patient can have multiple encounters, so we need to be careful about whether a KPI is encounter-level or patient-level.
weight has 96.86% unknown values.

In [16]:
mapping.info()

NameError: name 'mapping' is not defined

In [15]:
import os

print(os.getcwd())

E:\NIT Class\Data_analyst_projects


In [17]:
os.listdir()

['.ipynb_checkpoints',
 'customer_behavior_analysis.ipynb',
 'customer_shopping_behavior.csv',
 'diabetic_data.csv',
 'IDS_mapping.csv',
 'Patient-Readmission-Risk-Healthcare-Analytics.ipynb']

In [18]:
import pandas as pd

mapping = pd.read_csv("IDS_mapping.csv")

mapping.head(20)

,admission_type_id,description
0,1,Emergency
1,2,Urgent
2,3,Elective
3,4,Newborn
4,5,Not Available
5,6,NaN
6,7,Trauma Center
7,8,Not Mapped
8,NaN,NaN
9,discharge_disposition_id,description


In [19]:
mapping.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 67 entries, 0 to 66
Data columns (total 2 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   admission_type_id  65 non-null     object
 1   description        62 non-null     object
dtypes: object(2)
memory usage: 1.2+ KB


In [20]:
mapping.columns.tolist()

['admission_type_id', 'description']

In [21]:
mapping.head(30)

,admission_type_id,description
0,1,Emergency
1,2,Urgent
2,3,Elective
3,4,Newborn
4,5,Not Available
5,6,NaN
6,7,Trauma Center
7,8,Not Mapped
8,NaN,NaN
9,discharge_disposition_id,description


In [22]:
mapping.tail(30)

,admission_type_id,description
37,27,Discharged/transferred to a federal health car...
38,28,Discharged/transferred/referred to a psychiatr...
39,29,Discharged/transferred to a Critical Access Ho...
40,NaN,NaN
41,admission_source_id,description
42,1,Physician Referral
43,2,Clinic Referral
44,3,HMO Referral
45,4,Transfer from a hospital
46,5,Transfer from a Skilled Nursing Facility (SNF)


In [23]:
print(df['admission_type_id'].value_counts().sort_index())

admission_type_id
1    53990
2    18480
3    18869
4       10
5     4785
6     5291
7       21
8      320
Name: count, dtype: int64


In [24]:
print(df['discharge_disposition_id'].value_counts().sort_index())

discharge_disposition_id
1     60234
2      2128
3     13954
4       815
5      1184
6     12902
7       623
8       108
9        21
10        6
11     1642
12        3
13      399
14      372
15       63
16       11
17       14
18     3691
19        8
20        2
22     1993
23      412
24       48
25      989
27        5
28      139
Name: count, dtype: int64


In [25]:
print(df['admission_source_id'].value_counts().sort_index())

admission_source_id
1     29565
2      1104
3       187
4      3187
5       855
6      2264
7     57494
8        16
9       125
10        8
11        2
13        1
14        2
17     6781
20      161
22       12
25        2
Name: count, dtype: int64


DATA DICTIONARY

Purpose:
Understand each dataset column, its business meaning, and whether it should be
kept, transformed, or excluded before data cleaning.

In [26]:
# Display all columns with their data types

for col in df.columns:
    print(f"{col:30} {df[col].dtype}")

encounter_id                   int64
patient_nbr                    int64
race                           object
gender                         object
age                            object
weight                         object
admission_type_id              int64
discharge_disposition_id       int64
admission_source_id            int64
time_in_hospital               int64
payer_code                     object
medical_specialty              object
num_lab_procedures             int64
num_procedures                 int64
num_medications                int64
number_outpatient              int64
number_emergency               int64
number_inpatient               int64
diag_1                         object
diag_2                         object
diag_3                         object
number_diagnoses               int64
max_glu_serum                  object
A1Cresult                      object
metformin                      object
repaglinide                    object
nateglinide              

Now we know the structure of all 50 columns.
We are going to divide them into 3 groups:

🟢 KEEP → directly useful for our analysis
🟡 TRANSFORM → useful, but needs cleaning/renaming/grouping
🔴 EXCLUDE → not useful for this project's main analysis

🟢 KEEP	encounter_id, patient_nbr, race, gender, age, time_in_hospital, num_lab_procedures, num_procedures, num_medications, number_outpatient, number_emergency, number_inpatient, number_diagnoses, readmitted
🟡 TRANSFORM	admission_type_id, discharge_disposition_id, admission_source_id, payer_code, medical_specialty, diag_1, diag_2, diag_3, max_glu_serum, A1Cresult, change, diabetesMed, insulin
🔴 EXCLUDE	weight + the individual medication columns such as metformin, glipizide, glyburide, etc.

In [27]:
print("ADMISSION TYPE")
print(df["admission_type_id"].value_counts().sort_index())

print("\nDISCHARGE DISPOSITION")
print(df["discharge_disposition_id"].value_counts().sort_index())

print("\nADMISSION SOURCE")
print(df["admission_source_id"].value_counts().sort_index())

ADMISSION TYPE
admission_type_id
1    53990
2    18480
3    18869
4       10
5     4785
6     5291
7       21
8      320
Name: count, dtype: int64

DISCHARGE DISPOSITION
discharge_disposition_id
1     60234
2      2128
3     13954
4       815
5      1184
6     12902
7       623
8       108
9        21
10        6
11     1642
12        3
13      399
14      372
15       63
16       11
17       14
18     3691
19        8
20        2
22     1993
23      412
24       48
25      989
27        5
28      139
Name: count, dtype: int64

ADMISSION SOURCE
admission_source_id
1     29565
2      1104
3       187
4      3187
5       855
6      2264
7     57494
8        16
9       125
10        8
11        2
13        1
14        2
17     6781
20      161
22       12
25        2
Name: count, dtype: int64


In [28]:
mapping = pd.read_csv("IDS_mapping.csv")

print(mapping.shape)
print(mapping.columns.tolist())
display(mapping.head(20))

(67, 2)
['admission_type_id', 'description']


,admission_type_id,description
0,1,Emergency
1,2,Urgent
2,3,Elective
3,4,Newborn
4,5,Not Available
5,6,NaN
6,7,Trauma Center
7,8,Not Mapped
8,NaN,NaN
9,discharge_disposition_id,description


In [29]:
admission_type_map = {
    1: "Emergency",
    2: "Urgent",
    3: "Elective",
    4: "Newborn",
    5: "Not Available",
    6: "Unknown",
    7: "Trauma Center",
    8: "Not Mapped"
}

df["admission_type"] = df["admission_type_id"].map(admission_type_map)

df[["admission_type_id", "admission_type"]].drop_duplicates().sort_values("admission_type_id")

,admission_type_id,admission_type
1,1,Emergency
5,2,Urgent
6,3,Elective
2043,4,Newborn
3089,5,Not Available
0,6,Unknown
45829,7,Trauma Center
7789,8,Not Mapped


Map discharge_disposition_id

In [30]:
discharge_disposition_map = {
    1: "Discharged to home",
    2: "Transferred to short-term hospital",
    3: "Transferred to SNF",
    4: "Transferred to ICF",
    5: "Transferred to other inpatient facility",
    6: "Home with home health service",
    7: "Left AMA",
    8: "Home under Home IV provider",
    9: "Admitted as inpatient to this hospital",
    10: "Neonatal aftercare hospital",
    11: "Expired",
    12: "Still patient / outpatient return",
    13: "Hospice / home",
    14: "Hospice / medical facility",
    15: "Medicare swing bed",
    16: "Referred for outpatient services",
    17: "Referred to this institution for outpatient services",
    18: "Unknown",
    19: "Expired at home",
    20: "Expired in medical facility",
    21: "Expired / place unknown",
    22: "Transferred to rehab facility",
    23: "Transferred to long-term care hospital",
    24: "Transferred to Medicaid nursing facility",
    25: "Not Mapped",
    26: "Unknown / Invalid",
    27: "Transferred to federal healthcare facility",
    28: "Transferred to psychiatric facility",
    29: "Transferred to Critical Access Hospital",
    30: "Other healthcare institution"
}

df["discharge_disposition"] = df["discharge_disposition_id"].map(
    discharge_disposition_map
)

df[["discharge_disposition_id", "discharge_disposition"]]\
    .drop_duplicates()\
    .sort_values("discharge_disposition_id")

,discharge_disposition_id,discharge_disposition
1,1,Discharged to home
29,2,Transferred to short-term hospital
9,3,Transferred to SNF
972,4,Transferred to ICF
31,5,Transferred to other inpatient facility
13,6,Home with home health service
82,7,Left AMA
3558,8,Home under Home IV provider
16783,9,Admitted as inpatient to this hospital
487,10,Neonatal aftercare hospital


Map admission_source_id

In [31]:
admission_source_map = {
    1: "Physician Referral",
    2: "Clinic Referral",
    3: "HMO Referral",
    4: "Transfer from Hospital",
    5: "Transfer from SNF",
    6: "Transfer from Healthcare Facility",
    7: "Emergency Room",
    8: "Court / Law Enforcement",
    9: "Not Available",
    10: "Transfer from Critical Access Hospital",
    11: "Normal Delivery",
    12: "Premature Delivery",
    13: "Sick Baby",
    14: "Extramural Birth",
    15: "Not Available",
    17: "Unknown",
    18: "Transfer from Home Health Agency",
    19: "Readmission to Same Home Health Agency",
    20: "Not Mapped",
    21: "Unknown / Invalid",
    22: "Transfer from Hospital / Same Facility",
    23: "Born Inside Hospital",
    24: "Born Outside Hospital",
    25: "Transfer from Ambulatory Surgery Center",
    26: "Transfer from Hospice"
}

df["admission_source"] = df["admission_source_id"].map(
    admission_source_map
)

df[["admission_source_id", "admission_source"]]\
    .drop_duplicates()\
    .sort_values("admission_source_id")

,admission_source_id,admission_source
0,1,Physician Referral
5,2,Clinic Referral
1014,3,HMO Referral
8,4,Transfer from Hospital
335,5,Transfer from SNF
396,6,Transfer from Healthcare Facility
1,7,Emergency Room
1595,8,Court / Law Enforcement
24884,9,Not Available
44582,10,Transfer from Critical Access Hospital


In [33]:
admission_source_map = {
    1: "Physician Referral",
    2: "Clinic Referral",
    3: "HMO Referral",
    4: "Transfer from Hospital",
    5: "Transfer from SNF",
    6: "Transfer from Healthcare Facility",
    7: "Emergency Room",
    8: "Court / Law Enforcement",
    9: "Not Available",
    10: "Transfer from Critical Access Hospital",
    11: "Normal Delivery",
    12: "Premature Delivery",
    13: "Sick Baby",
    14: "Extramural Birth",
    15: "Not Available",
    17: "Unknown",
    18: "Transfer from Home Health Agency",
    19: "Readmission to Same Home Health Agency",
    20: "Not Mapped",
    21: "Unknown / Invalid",
    22: "Transfer from Hospital / Same Facility",
    23: "Born Inside Hospital",
    24: "Born Outside Hospital",
    25: "Transfer from Ambulatory Surgery Center",
    26: "Transfer from Hospice"
}

df["admission_source"] = df["admission_source_id"].map(
    admission_source_map
)

df[["admission_source_id", "admission_source"]] \
    .drop_duplicates() \
    .sort_values("admission_source_id")

,admission_source_id,admission_source
0,1,Physician Referral
5,2,Clinic Referral
1014,3,HMO Referral
8,4,Transfer from Hospital
335,5,Transfer from SNF
396,6,Transfer from Healthcare Facility
1,7,Emergency Room
1595,8,Court / Law Enforcement
24884,9,Not Available
44582,10,Transfer from Critical Access Hospital


Now we check missing values

In [35]:
missing_counts = (df == "?").sum()

missing_counts[missing_counts > 0].sort_values(ascending=False)

weight               98569
medical_specialty    49949
payer_code           40256
race                  2273
diag_3                1423
diag_2                 358
diag_1                  21
dtype: int64

approach

We will not simply delete rows containing ?.

Instead:

weight → likely exclude from our project because almost the entire column is missing.
medical_specialty → keep, convert ? → "Unknown".
payer_code → keep/transform, convert ? → "Unknown".
race → convert ? → "Unknown".
diag_1, diag_2, diag_3 → keep, handle ? as "Unknown".
Why?

If we deleted every row containing ?, we'd unnecessarily lose a large amount of useful patient data.

For example, weight has:

98,569 / 101,766 ≈ 96.9% missing.

So there's very little value in keeping it for this project.

In [36]:
categorical_columns = [
    "race",
    "gender",
    "age",
    "payer_code",
    "medical_specialty",
    "max_glu_serum",
    "A1Cresult",
    "change",
    "diabetesMed",
    "readmitted"
]

for col in categorical_columns:
    print(f"\n{'='*50}")
    print(f"{col}")
    print(f"{'='*50}")
    print(df[col].value_counts(dropna=False))


race
race
Caucasian          76099
AfricanAmerican    19210
?                   2273
Hispanic            2037
Other               1506
Asian                641
Name: count, dtype: int64

gender
gender
Female             54708
Male               47055
Unknown/Invalid        3
Name: count, dtype: int64

age
age
[70-80)     26068
[60-70)     22483
[50-60)     17256
[80-90)     17197
[40-50)      9685
[30-40)      3775
[90-100)     2793
[20-30)      1657
[10-20)       691
[0-10)        161
Name: count, dtype: int64

payer_code
payer_code
?     40256
MC    32439
HM     6274
SP     5007
BC     4655
MD     3532
CP     2533
UN     2448
CM     1937
OG     1033
PO      592
DM      549
CH      146
WC      135
OT       95
MP       79
SI       55
FR        1
Name: count, dtype: int64

medical_specialty
medical_specialty
?                         49949
InternalMedicine          14635
Emergency/Trauma           7565
Family/GeneralPractice     7440
Cardiology                 5352
                    

In [37]:
for col in ["diag_1", "diag_2", "diag_3"]:
    print(f"\n{col}")
    print(df[col].value_counts(dropna=False).head(20))


diag_1
diag_1
428      6862
414      6581
786      4016
410      3614
486      3508
427      2766
491      2275
715      2151
682      2042
434      2028
780      2019
996      1967
276      1889
38       1688
250.8    1680
599      1595
584      1520
V57      1207
250.6    1183
518      1115
Name: count, dtype: int64

diag_2
diag_2
276       6752
428       6662
250       6071
427       5036
401       3736
496       3305
599       3288
403       2823
414       2650
411       2566
250.02    2074
707       1999
585       1871
584       1649
491       1545
250.01    1523
285       1520
780       1491
425       1434
682       1433
Name: count, dtype: int64

diag_3
diag_3
250       11555
401        8289
276        5175
428        4577
427        3955
414        3664
496        2605
403        2357
585        1992
272        1969
599        1941
?          1423
V45        1389
250.02     1369
707        1360
780        1334
285        1200
425        1136
250.6      1080
424        1063
Nam

In [38]:
medication_cols = [
    "metformin",
    "repaglinide",
    "nateglinide",
    "chlorpropamide",
    "glimepiride",
    "acetohexamide",
    "glipizide",
    "glyburide",
    "tolbutamide",
    "pioglitazone",
    "rosiglitazone",
    "acarbose",
    "miglitol",
    "troglitazone",
    "tolazamide",
    "examide",
    "citoglipton",
    "insulin",
    "glyburide-metformin",
    "glipizide-metformin",
    "glimepiride-pioglitazone",
    "metformin-rosiglitazone",
    "metformin-pioglitazone"
]

for col in medication_cols:
    print(f"\n{col}")
    print(df[col].value_counts(dropna=False))


metformin
metformin
No        81778
Steady    18346
Up         1067
Down        575
Name: count, dtype: int64

repaglinide
repaglinide
No        100227
Steady      1384
Up           110
Down          45
Name: count, dtype: int64

nateglinide
nateglinide
No        101063
Steady       668
Up            24
Down          11
Name: count, dtype: int64

chlorpropamide
chlorpropamide
No        101680
Steady        79
Up             6
Down           1
Name: count, dtype: int64

glimepiride
glimepiride
No        96575
Steady     4670
Up          327
Down        194
Name: count, dtype: int64

acetohexamide
acetohexamide
No        101765
Steady         1
Name: count, dtype: int64

glipizide
glipizide
No        89080
Steady    11356
Up          770
Down        560
Name: count, dtype: int64

glyburide
glyburide
No        91116
Steady     9274
Up          812
Down        564
Name: count, dtype: int64

tolbutamide
tolbutamide
No        101743
Steady        23
Name: count, dtype: int64

pioglitazone
p

In [39]:
numeric_cols = [
    "time_in_hospital",
    "num_lab_procedures",
    "num_procedures",
    "num_medications",
    "number_outpatient",
    "number_emergency",
    "number_inpatient",
    "number_diagnoses"
]

df[numeric_cols].describe().T

,count,mean,std,min,25%,50%,75%,max
time_in_hospital,101766.0,4.395987,2.985108,1.0,2.0,4.0,6.0,14.0
num_lab_procedures,101766.0,43.095641,19.674362,1.0,31.0,44.0,57.0,132.0
num_procedures,101766.0,1.339730,1.705807,0.0,0.0,1.0,2.0,6.0
num_medications,101766.0,16.021844,8.127566,1.0,10.0,15.0,20.0,81.0
number_outpatient,101766.0,0.369357,1.267265,0.0,0.0,0.0,0.0,42.0
number_emergency,101766.0,0.197836,0.930472,0.0,0.0,0.0,0.0,76.0
number_inpatient,101766.0,0.635566,1.262863,0.0,0.0,0.0,1.0,21.0
number_diagnoses,101766.0,7.422607,1.933600,1.0,6.0,8.0,9.0,16.0


In [40]:
df["number_diagnoses"].describe()

count    101766.000000
mean          7.422607
std           1.933600
min           1.000000
25%           6.000000
50%           8.000000
75%           9.000000
max          16.000000
Name: number_diagnoses, dtype: float64

In [41]:
print("Unique values:", df["number_diagnoses"].nunique())
print("Maximum:", df["number_diagnoses"].max())
print("Top values:")
print(df["number_diagnoses"].value_counts().sort_index())

Unique values: 16
Maximum: 16
Top values:
number_diagnoses
1       219
2      1023
3      2835
4      5537
5     11393
6     10161
7     10393
8     10616
9     49474
10       17
11       11
12        9
13       16
14        7
15       10
16       45
Name: count, dtype: int64


In [42]:
df_clean = df.copy()

print(df_clean.shape)

(101766, 53)


In [44]:
unknown_cols = [
    "race",
    "payer_code",
    "medical_specialty",
    "diag_1",
    "diag_2",
    "diag_3"
]

for col in unknown_cols:
    df_clean[col] = df_clean[col].replace("?", "Unknown")

In [45]:
for col in unknown_cols:
    print(f"{col}: {(df_clean[col] == '?').sum()} remaining")

race: 0 remaining
payer_code: 0 remaining
medical_specialty: 0 remaining
diag_1: 0 remaining
diag_2: 0 remaining
diag_3: 0 remaining


In [46]:
df_clean = df_clean.drop(columns=["weight"])

print(df_clean.shape)

(101766, 52)


In [47]:
df_clean["max_glu_serum"] = df_clean["max_glu_serum"].fillna("Not Tested")
df_clean["A1Cresult"] = df_clean["A1Cresult"].fillna("Not Tested")

In [48]:
print(df_clean["max_glu_serum"].value_counts())
print()
print(df_clean["A1Cresult"].value_counts())

max_glu_serum
Not Tested    96420
Norm           2597
>200           1485
>300           1264
Name: count, dtype: int64

A1Cresult
Not Tested    84748
>8             8216
Norm           4990
>7             3812
Name: count, dtype: int64


creating 30 days readmission target 

In [50]:
df_clean["readmitted_30_days"] = df_clean["readmitted"].map({
    "<30": "Yes",
    ">30": "No",
    "NO": "No"
})

In [51]:
print(df_clean["readmitted_30_days"].value_counts(dropna=False))

readmitted_30_days
No     90409
Yes    11357
Name: count, dtype: int64


In [52]:
df_clean["gender"] = df_clean["gender"].replace(
    "Unknown/Invalid", "Unknown"
)

print(df_clean["gender"].value_counts())

gender
Female     54708
Male       47055
Unknown        3
Name: count, dtype: int64


In [53]:
print(df_clean["race"].value_counts())

race
Caucasian          76099
AfricanAmerican    19210
Unknown             2273
Hispanic            2037
Other               1506
Asian                641
Name: count, dtype: int64


In [54]:
print(df_clean["payer_code"].value_counts())

payer_code
Unknown    40256
MC         32439
HM          6274
SP          5007
BC          4655
MD          3532
CP          2533
UN          2448
CM          1937
OG          1033
PO           592
DM           549
CH           146
WC           135
OT            95
MP            79
SI            55
FR             1
Name: count, dtype: int64


In [55]:
print("Unique specialties:", df_clean["medical_specialty"].nunique())
print()
print(df_clean["medical_specialty"].value_counts().head(20))

Unique specialties: 73

medical_specialty
Unknown                              49949
InternalMedicine                     14635
Emergency/Trauma                      7565
Family/GeneralPractice                7440
Cardiology                            5352
Surgery-General                       3099
Nephrology                            1613
Orthopedics                           1400
Orthopedics-Reconstructive            1233
Radiologist                           1140
Pulmonology                            871
Psychiatry                             854
Urology                                685
ObstetricsandGynecology                671
Surgery-Cardiovascular/Thoracic        652
Gastroenterology                       564
Surgery-Vascular                       533
Surgery-Neuro                          468
PhysicalMedicineandRehabilitation      391
Oncology                               348
Name: count, dtype: int64


In [56]:
diagnosis_cols = ["diag_1", "diag_2", "diag_3"]

for col in diagnosis_cols:
    print(f"\n{col}")
    print("Unique values:", df_clean[col].nunique())
    print("Unknown:", (df_clean[col] == "Unknown").sum())


diag_1
Unique values: 717
Unknown: 21

diag_2
Unique values: 749
Unknown: 358

diag_3
Unique values: 790
Unknown: 1423


In [57]:
df_clean["age_group"] = (
    df_clean["age"]
    .str.replace("[", "", regex=False)
    .str.replace(")", "", regex=False)
)

print(df_clean["age_group"].value_counts().sort_index())

age_group
0-10        161
10-20       691
20-30      1657
30-40      3775
40-50      9685
50-60     17256
60-70     22483
70-80     26068
80-90     17197
90-100     2793
Name: count, dtype: int64


In [58]:
import pandas as pd

# ============================================================
# 1. Create a clean copy
# ============================================================

df_clean = df.copy()


# ============================================================
# 2. Replace '?' with 'Unknown'
# ============================================================

unknown_cols = [
    "race",
    "payer_code",
    "medical_specialty",
    "diag_1",
    "diag_2",
    "diag_3"
]

for col in unknown_cols:
    df_clean[col] = df_clean[col].replace("?", "Unknown")


# ============================================================
# 3. Standardize gender
# ============================================================

df_clean["gender"] = df_clean["gender"].replace(
    "Unknown/Invalid", "Unknown"
)


# ============================================================
# 4. Remove weight
#    ~97% values are missing
# ============================================================

df_clean = df_clean.drop(columns=["weight"])


# ============================================================
# 5. Handle glucose and A1C missing values
# ============================================================

df_clean["max_glu_serum"] = df_clean["max_glu_serum"].fillna(
    "Not Tested"
)

df_clean["A1Cresult"] = df_clean["A1Cresult"].fillna(
    "Not Tested"
)


# ============================================================
# 6. Create 30-day readmission target
# ============================================================

df_clean["readmitted_30_days"] = df_clean["readmitted"].map({
    "<30": "Yes",
    ">30": "No",
    "NO": "No"
})


# ============================================================
# 7. Create clean age group
# ============================================================

df_clean["age_group"] = (
    df_clean["age"]
    .str.replace("[", "", regex=False)
    .str.replace(")", "", regex=False)
)


# ============================================================
# 8. Create Length of Stay groups
# ============================================================

df_clean["los_group"] = pd.cut(
    df_clean["time_in_hospital"],
    bins=[0, 2, 4, 7, 14],
    labels=[
        "1-2 days",
        "3-4 days",
        "5-7 days",
        "8-14 days"
    ],
    include_lowest=True
)


# ============================================================
# 9. Create outpatient utilization group
# ============================================================

df_clean["outpatient_visit_group"] = pd.cut(
    df_clean["number_outpatient"],
    bins=[-1, 0, 2, float("inf")],
    labels=[
        "0 visits",
        "1-2 visits",
        "3+ visits"
    ]
)


# ============================================================
# 10. Create emergency utilization group
# ============================================================

df_clean["emergency_visit_group"] = pd.cut(
    df_clean["number_emergency"],
    bins=[-1, 0, 1, float("inf")],
    labels=[
        "0 visits",
        "1 visit",
        "2+ visits"
    ]
)


# ============================================================
# 11. Create inpatient utilization group
# ============================================================

df_clean["inpatient_visit_group"] = pd.cut(
    df_clean["number_inpatient"],
    bins=[-1, 0, 1, float("inf")],
    labels=[
        "0 visits",
        "1 visit",
        "2+ visits"
    ]
)


# ============================================================
# 12. Remove extremely sparse medication columns
# ============================================================

medications_to_remove = [
    "repaglinide",
    "nateglinide",
    "chlorpropamide",
    "glimepiride",
    "acetohexamide",
    "glipizide",
    "glyburide",
    "tolbutamide",
    "pioglitazone",
    "rosiglitazone",
    "acarbose",
    "miglitol",
    "troglitazone",
    "tolazamide",
    "examide",
    "citoglipton",
    "glyburide-metformin",
    "glipizide-metformin",
    "glimepiride-pioglitazone",
    "metformin-rosiglitazone",
    "metformin-pioglitazone"
]

df_clean = df_clean.drop(
    columns=medications_to_remove
)


# ============================================================
# 13. Standardize column names
# ============================================================

df_clean.columns = (
    df_clean.columns
    .str.lower()
    .str.strip()
    .str.replace("-", "_", regex=False)
    .str.replace(" ", "_", regex=False)
)


# ============================================================
# 14. Final validation
# ============================================================

print("Shape:", df_clean.shape)

print("\nDuplicate rows:", df_clean.duplicated().sum())

print("\nMissing values:")
print(df_clean.isna().sum()[df_clean.isna().sum() > 0])

print("\n30-day readmission:")
print(df_clean["readmitted_30_days"].value_counts())

print("\nAge groups:")
print(df_clean["age_group"].value_counts().sort_index())

print("\nLength of stay groups:")
print(df_clean["los_group"].value_counts().sort_index())

print("\nFirst 5 rows:")
display(df_clean.head())

Shape: (101766, 37)

Duplicate rows: 0

Missing values:
Series([], dtype: int64)

30-day readmission:
readmitted_30_days
No     90409
Yes    11357
Name: count, dtype: int64

Age groups:
age_group
0-10        161
10-20       691
20-30      1657
30-40      3775
40-50      9685
50-60     17256
60-70     22483
70-80     26068
80-90     17197
90-100     2793
Name: count, dtype: int64

Length of stay groups:
los_group
1-2 days     31432
3-4 days     31680
5-7 days     23364
8-14 days    15290
Name: count, dtype: int64

First 5 rows:


,encounter_id,patient_nbr,race,gender,age,admission_type_id,discharge_disposition_id,admission_source_id,time_in_hospital,payer_code,...,readmitted,admission_type,discharge_disposition,admission_source,readmitted_30_days,age_group,los_group,outpatient_visit_group,emergency_visit_group,inpatient_visit_group
0,2278392,8222157,Caucasian,Female,[0-10),6,25,1,1,Unknown,...,NO,Unknown,Not Mapped,Physician Referral,No,0-10,1-2 days,0 visits,0 visits,0 visits
1,149190,55629189,Caucasian,Female,[10-20),1,1,7,3,Unknown,...,>30,Emergency,Discharged to home,Emergency Room,No,10-20,3-4 days,0 visits,0 visits,0 visits
2,64410,86047875,AfricanAmerican,Female,[20-30),1,1,7,2,Unknown,...,NO,Emergency,Discharged to home,Emergency Room,No,20-30,1-2 days,1-2 visits,0 visits,1 visit
3,500364,82442376,Caucasian,Male,[30-40),1,1,7,2,Unknown,...,NO,Emergency,Discharged to home,Emergency Room,No,30-40,1-2 days,0 visits,0 visits,0 visits
4,16680,42519267,Caucasian,Male,[40-50),1,1,7,1,Unknown,...,NO,Emergency,Discharged to home,Emergency Room,No,40-50,1-2 days,0 visits,0 visits,0 visits


In [59]:
# ============================================================
# FINAL DATA VALIDATION
# ============================================================

print("1. Dataset shape:")
print(df_clean.shape)

print("\n2. Duplicate rows:")
print(df_clean.duplicated().sum())

print("\n3. Duplicate encounter IDs:")
print(df_clean["encounter_id"].duplicated().sum())

print("\n4. Missing values:")
missing = df_clean.isna().sum()
print(missing[missing > 0])

print("\n5. 30-day readmission:")
print(df_clean["readmitted_30_days"].value_counts())

print("\n6. Gender:")
print(df_clean["gender"].value_counts())

print("\n7. Race:")
print(df_clean["race"].value_counts())

print("\n8. Age groups:")
print(df_clean["age_group"].value_counts().sort_index())

print("\n9. Length of stay:")
print(df_clean["los_group"].value_counts().sort_index())

print("\n10. Data types:")
print(df_clean.dtypes)

1. Dataset shape:
(101766, 37)

2. Duplicate rows:
0

3. Duplicate encounter IDs:
0

4. Missing values:
Series([], dtype: int64)

5. 30-day readmission:
readmitted_30_days
No     90409
Yes    11357
Name: count, dtype: int64

6. Gender:
gender
Female     54708
Male       47055
Unknown        3
Name: count, dtype: int64

7. Race:
race
Caucasian          76099
AfricanAmerican    19210
Unknown             2273
Hispanic            2037
Other               1506
Asian                641
Name: count, dtype: int64

8. Age groups:
age_group
0-10        161
10-20       691
20-30      1657
30-40      3775
40-50      9685
50-60     17256
60-70     22483
70-80     26068
80-90     17197
90-100     2793
Name: count, dtype: int64

9. Length of stay:
los_group
1-2 days     31432
3-4 days     31680
5-7 days     23364
8-14 days    15290
Name: count, dtype: int64

10. Data types:
encounter_id                   int64
patient_nbr                    int64
race                          object
gender           

cleaning is completed now 

In [60]:
# ============================================================
# EXPORT CLEANED DATASET
# ============================================================

# Convert categorical columns to string/object
# so PostgreSQL imports them cleanly
category_cols = [
    "los_group",
    "outpatient_visit_group",
    "emergency_visit_group",
    "inpatient_visit_group"
]

for col in category_cols:
    df_clean[col] = df_clean[col].astype(str)

# Save cleaned dataset
output_path = "healthcare_readmission_cleaned.csv"

df_clean.to_csv(
    output_path,
    index=False
)

print("Cleaned dataset exported successfully.")
print("File:", output_path)
print("Shape:", df_clean.shape)

Cleaned dataset exported successfully.
File: healthcare_readmission_cleaned.csv
Shape: (101766, 37)
